ls1intum/Artemis
# `Programming exercises`: Add general structure for code hints

### Checklist
#### General
(https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA).


### Description
This PR introduces the general structure of the code hints corresponding to the below Database Model. The automatic generation is not part of this PR and will be introduced in later follow-up PRs.

![Database Model](https://user-images.githubusercontent.com/73833780/152537210-7c90ebe9-6ca3-423f-ae98-5fe513699721.png)

 A `ProgrammingExercise` already aggregates multiple `ProgrammingExerciseTestCase`s into tasks, but only in the textual form of the problem statement. In order to provide high-quality `CodeHint`s, this PR extracts the tasks from the problem statement and convert them to  `ProgrammingExerciseTask`s so we can create one `CodeHint` per `ProgrammingExerciseTask` and not per `ProgrammingExerciseTestCase`. 
`CodeHint`s aggregate the `SolutionEntries` of all `ProgrammingExerciseTestCase`s of their  `ProgrammingExerciseTask`. This results in fewer `ExerciseHint`s per `ProgrammingExercise` which increases their usability.  One `SolutionEntry` represents a single change in a single file.


### New endpoints
ProgrammingExerciseTaskResource
-	GET `/api/programming-exercises/{exerciseId}/tasks`: Return all tasks for the programming exercise with exerciseId. (Student)

ProgrammingExerciseSolutionEntryResource
-	GET `/api/programming-exercises/{exerciseId}/solution-entries/{solutionEntryId}`: Return the solutionEntry with solutionEntryId. (Tutor)
-	GET `/api/programming-exercises/{exerciseId}/code-hints/{codeHintId}/solution-entries`: Return all solution entries for a code hint of a programming exercise. Will be used for client-side rendering of the code hint when opening a code hint. (Student)
-	GET `/api/programming-exercises/{exerciseId}/test-cases/{testCaseId}/solution-entries`: Return all solution entries for a test case of a programming exercise. Will be used when displaying a solution entry when editing it. (Student)
-	POST `/api/programming-exercises/{exerciseId}/test-cases/{testCaseId}/solution-entries`: Create a solution entry for a test case of a programming exercise. Will be used to allow the instructor to manually create solution entries for a test case. (Editor)
-	PUT `/api/programming-exercises/{exerciseId}/test-cases/{testCaseId}/solution-entries/{solutionEntryId}`: Update a solution entry for a test case of a programming exercise. Will be used to allow the instructor to manually update a solution entry for a test case. (Editor)
-	DELETE `/api/programming-exercises/{exerciseId}/test-cases/{testCaseId}/solution-entries/{solutionEntryId}`: Delete a solution entry for a test case of a programming exercise. Will be used to allow the instructor to manually delete a solution entry for a test case. (Editor)

ProgrammingExerciseResource
-	DELETE `/api/programming-exercises/{exerciseId}/tasks`: Deletes all tasks and all solution entries for a programming exercise. Will be used to reset the mapping in case of not considered edge cases. (Editor)

### New Buttons on Client
This PR introduces two new buttons on the edit page for a programming exercise to allow a more user friendly testing (only visible as admin):
1. "Load extracted tasks": Retrieve and display all tasks for this programming exercise. (Note: this button does not trigger the extraction anymore but only loads existing tasks. After pressing button 2, therefore 0 tasks should be displayed.
2. "Delete tasks": Invoke deleting all tasks and solution entries in order to recover from errors due to possible unconsidered edge case.


### Steps for Testing
#### Requirements:
- 1 Admin (artemis_test_user_12 on TS3 works)
- 1 programming exercise with at least one hint that was added before the PR was deployed
- Note: The PR can currently only be tested on TS2 because Bitbucket is down on TS3 and there are no admin accounts for testing on TS1.

Verify the following as an instructor:
1. Check that hints are still working
    - Hints for existing programming exercises are still working (existing ones can be displayed, new ones can be created/updated/deleted)
    - Hints for newly created programming exercises are working (new hints can be created/updated/deleted)
2. Check that the loading and displaying the extracted tasks work.
    - When creating a programming exercise
    - When importing a programming exercise
    - When updating the problem statement of a programming exercise
    - When updating the test cases in the test case repo
    - When pressing the specific button
3. After pressing the "Delete tasks"-Button, and clicking on the "Load extracted tasks"-Button, 0 tasks should be displayed.


### Review Progress

#### Code Review
#### Manual Tests

### Screenshots
<img width="596" alt="Bildschirmfoto 2022-02-21 um 17 07 15" src="https://user-images.githubusercontent.com/73833780/154991181-b47a02c2-2565-419d-92cb-648c4d3d29ae.png">
